### ISSUE TITLE
Fetch url escaped on iOS 17

### ISSUE THREAD

**Issue by Reeywhaar** [facebook/react-native#39793]
### Description

When fetching something on iOS 17 with fetch url gets escaped. This is unexpected and differs from android. I can't confirm whether is it happens only in simulator and not on real device.

On the image below I created test reproduction. I have android and ios simulator and echo server that runs on localhost:8000. When tapping "Test fetch" button it simply calls endpoint on localhost server. As you can see server receives different request urls. Query on iOS gets escaped which is undesirable.

<img width="1310" alt="Screenshot 2023-10-04 at 01 58 46" src="https://github.com/facebook/react-native/assets/884649/da28395e-a3bc-4797-8f8d-671cdfb52649">

### React Native Version

0.72.5

### Output of `npx react-native info`

info Fetching system and libraries information...
System:
  OS: macOS 13.6
  CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
  Memory: 441.45 MB / 32.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 20.6.0
    path: /usr/local/bin/node
  Yarn:
    version: 1.22.4
    path: /usr/local/bin/yarn
  npm:
    version: 9.8.1
    path: /usr/local/bin/npm
  Watchman:
    version: 2023.09.04.00
    path: /usr/local/bin/watchman
Managers:
  CocoaPods:
    version: 1.12.1
    path: /usr/local/bin/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 23.0
      - iOS 17.0
      - macOS 14.0
      - tvOS 17.0
      - watchOS 10.0
  Android SDK:
    API Levels:
      - "23"
      - "28"
      - "29"
      - "30"
      - "31"
      - "33"
    Build Tools:
      - 30.0.2
      - 30.0.3
      - 31.0.0
      - 33.0.0
      - 33.0.2
      - 34.0.0
    System Images:
      - android-29 | Google Play Intel x86 Atom
      - android-30 | Google APIs Intel x86 Atom
      - android-30 | Google Play Intel x86 Atom
      - android-31 | Google APIs Intel x86_64 Atom
    Android NDK: Not Found
IDEs:
  Android Studio: 2022.2 AI-222.4459.24.2221.10121639
  Xcode:
    version: 15.0/15A240d
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 15.0.1
    path: /usr/bin/javac
  Ruby:
    version: 3.2.2
    path: /usr/local/opt/ruby/bin/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.2.0
    wanted: 18.2.0
  react-native:
    installed: 0.72.5
    wanted: 0.72.5
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: true
  newArchEnabled: false

### Steps to reproduce

I've added test reproduction repo:
https://github.com/Reeywhaar/react-native-fetch-bug

- Run `sh server.sh` to start echo server
- Run ios and android versions
- Push `Test fetch` button
- Check server logs

### Snack, screenshot, or link to a repository

https://github.com/Reeywhaar/react-native-fetch-bug

**Comment by tandaocong** [comment]
You must be "stringifying" the query param when calling API on iOS 17
```javascript
import qs from 'qs';

const query = qs.stringify({query: {id: [1, 2, 3]});
const url = `${your_domain}?${query}`;
``` 
See change log: https://developer.apple.com/documentation/foundation/nsurl/1572047-urlwithstring

**Comment by Reeywhaar** [comment]
Yes, problem appeared when I tried to stringify url with `query-string` and request url become wrong. But only on iOS

```ts
import qs from 'qs';

const query = qs.stringify({"query[id]": [1,2,3]});
const url = `${your_domain}?${query}`;
```

**Comment by tandaocong** [comment]
> Yes, problem appeared when I tried to stringify url with `query-string` and request url become wrong. But only on iOS
> 
> ```ts
> import qs from 'query-string';
> 
> const query = qs.stringify({"query[id]": [1,2,3]});
> 
> const url = `${your_domain}?${query}`;
> 
> console.log(url) // printed url is fine
> fetch(url)
> 
> 
> // server receives double escaped url
> ```
Do not use "query[id]" because the [ ] character cannot be escaped, you can const query = qs.stringify({query: {id: '1, 2, 3'}});

**Comment by Reeywhaar** [comment]
`query[id]` is perfectly fine parameter as well as `query[]`. Also, I repeat, it works fine on Android. Also I get different urls when printing it in console in the one I get on server. On server it gets double escaped

**Comment by Reeywhaar** [comment]
In any way, server should receive same url I put into `fetch`. But this is not what happens on iOS. `fetch` escapes url on iOS but should not do that

**Comment by soutua** [comment]
Also encountered this issue. Looks like it's possible to workaround the problem by not using the shorthand encoding like `+` for the spaces, but instead escaping all the special characters with the `%` syntax (e.g. `%20` instead of `+`), then iOS 17 seems to leave the url alone.

**Comment by devon94** [comment]
Looks like it was fixed here https://github.com/facebook/react-native/commit/9841bd81852d59608fe3566b17831d6d42eb7dcf but then reverted shortly after here https://github.com/facebook/react-native/commit/2be409ff55b50563a5f123f7823fa7cdb72dbef9 but for some internal reason?

I had a similar fix here that did seem to work in our case https://github.com/ArchonSystemsInc/react-native/commit/00fd6f5e1d6817ff8d962f9f3d1a6cc127049c76

**Comment by nucleartux** [comment]
Has anyone found a solution? The solutions mentioned above don’t work for me.

**Comment by nucleartux** [comment]
So, if anyone is curious, I made a really dirty hack. It's not tested in production, and I'm not an iOS developer, but it works for me. So, use it with caution.

```diff
diff --git a/React/Base/RCTConvert.mm b/React/Base/RCTConvert.mm
index b868ee72f65ce85b5a4ee2bb9725ce01ebf88739..7fae01a6d51ed8fc392e1f1f922934a44b7b4e6c 100644
--- a/React/Base/RCTConvert.mm
+++ b/React/Base/RCTConvert.mm
@@ -86,6 +86,12 @@ + (NSURL *)NSURL:(id)json
   @try { // NSURL has a history of crashing with bad input, so let's be safe
     NSURL *URL = [NSURL URLWithString:path];
     if (URL.scheme) { // Was a well-formed absolute URL
+      NSURLComponents *components = [NSURLComponents componentsWithString:path];
+      components.percentEncodedQuery = [components.percentEncodedQuery stringByReplacingOccurrencesOfString:@"%252B" withString:@"%2B"];
+      if (components.URL) {
+        return components.URL;
+      }    
+
       return URL;
     }
 
```

If you use Yarn, you can add it to your project by adding this line to the resolutions field, like so:
```json
"react-native@0.74.1": "patch:react-native@npm:0.74.1#.yarn/patches/react-native-npm-0.74.1-754c02ae9e.patch"
```
And run `yarn` command

**Comment by mattmatters** [comment]
This is killing us too since we do lots of nested arrays in get requests (filtering parameters).

**Comment by brianw-tech** [comment]
For anyone experiencing this issue, after some vibe coding, AI recommended adding this to my stringified JSON

`let _query = qs.stringify(
   .........
    {
      addQueryPrefix: true,
      encodeValuesOnly: true,
      format: 'RFC1738', // Ensures square brackets are encoded
    },
  );`

then
`// Replace encoded colon (%3A) with a plain colon (:)
  _query = _query.replace(/%3A/g, ':');`

**Comment by jayant-lumino** [comment]
Having the same issue 
On Android working fine

`[api/sliders?filter[country_id]=48&lang=en&api_version=2024-12-01](https://api.cartobike.com/api/sliders?filter[country_id]=48&lang=en&api_version=2024-12-01)`

But on ios it's encoding

[/api/sliders?filter%5Bcountry_id%5D=2&lang=en&api_version=2024-12-01](https://api.cartobike.com/api/sliders?filter%5Bcountry_id%5D=2&lang=en&api_version=2024-12-01)
